jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java

```diff
@@ -0,0 +1,156 @@
+    private int lengthBound;
+
+    /**
+     * Constructs an algorithm with given inputs
+     *
+     * @param graph the (improvement) graph on which to calculate the local augmentation
+     * @param lengthBound the upper bound for the length of cycles to detect
+     * @param labels the labels of the vertices encoding the subsets of vertices
+     */
+    public AhujaOrlinSharmaCyclicExchangeLocalAugmentation(Graph<V, E> graph, int lengthBound, Map<V, Integer> labels) {
+        this.graph = graph;
+        this.lengthBound = lengthBound;
+        this.labels = labels;
+    }
+
+    /**
+     * calculates a valid subset-disjoint negative cycle
+     *
+     * @return a valid subset-disjoint negative cycle encoded as LabeledPath
+     */
+    public LabeledPath<V> getLocalAugmentationCycle() {
+
+        int k = 1;
+        LabeledPath<V> C = new LabeledPath<>();
+
+        Set<LabeledPath<V>> PathsLengthK = new LinkedHashSet<>();
+        Set<LabeledPath<V>> PathsLengthKplus1 = new LinkedHashSet<>();
+
+        // initialize PathsLengthK for k = 1
+        for(E e : graph.edgeSet()) {
+            if(graph.getEdgeWeight(e) < 0) {
+                // initialize all paths of cost < 0
+                LinkedList<V> pathVertices = new LinkedList<>();
+                Map<V, Integer> pathLabels = new LinkedHashMap<>();
+                V sourceVertex = graph.getEdgeSource(e);
+                V targetVertex = graph.getEdgeTarget(e);
+                pathVertices.add(sourceVertex);
+                pathVertices.add(targetVertex);
+                pathLabels.put(sourceVertex, labels.get(sourceVertex));
+                pathLabels.put(targetVertex, labels.get(targetVertex));
+                LabeledPath<V> path = new LabeledPath<>(pathVertices, graph.getEdgeWeight(e), pathLabels);
+
+                // add path to set of paths of length 1
+                PathsLengthK.add(path);
+            }
+        }
+
+        while(k < lengthBound && C.getCost() >= 0) {
+            while(!PathsLengthK.isEmpty()) {
+                for(Iterator<LabeledPath<V>> it = PathsLengthK.iterator(); it.hasNext();) { // go through all valid paths of length k
+                    LabeledPath<V> path = it.next();
+                    it.remove();
+
+                    V head = path.getHead();
+                    V tail = path.getTail();
+
+                    if(graph.containsEdge(tail, head) && path.getCost() + graph.getEdgeWeight(graph.getEdge(tail, head)) < C.getCost()) { // the path builds a valid cycle
+                        C = path.clone();
+                        C.addVertex(head, graph.getEdgeWeight(graph.getEdge(tail, head)), labels.get(head));
+                        if(C.getCost() < 0) { // only return the cycle if it is negative (in the first iteration it can be non-negative)
+                            return C;
+                        }
+                    }
+
+                    for(E e : graph.outgoingEdgesOf(tail)) {
```

Comment:
Is this equivalent to the algorithm in the lecture notes provided? It seems like head/tail are switched.

Reply:
no, I switched head and tail as I consider the paths more like queue.